include/pybind11/pybind11.h

```diff
@@ -1478,74 +1514,6 @@ class module_ : public object {
     }
-
-    /// Must be a POD type, and must hold enough entries for all of the possible slots PLUS ONE for
-    /// the sentinel (0) end slot.
-    using slots_array = std::array<PyModuleDef_Slot, 4>;
-
-    /** \rst
-        Initialize a module def for use with multi-phase module initialization.
-
-        ``def`` should point to a statically allocated module_def.
-        ``slots`` must already contain a Py_mod_exec or Py_mod_create slot and will be filled with
-            additional slots from the supplied options (and the empty sentinel slot).
-    \endrst */
-    template <typename... Options>
-    static object initialize_multiphase_module_def(const char *name,
-                                                   const char *doc,
-                                                   module_def *def,
-                                                   slots_array &slots,
-                                                   Options &&...options) {
-        size_t next_slot = 0;
-        size_t term_slot = slots.size() - 1;
-
-        // find the end of the supplied slots
-        while (next_slot < term_slot && slots[next_slot].slot != 0) {
-            ++next_slot;
-        }
-
-#ifdef Py_mod_multiple_interpreters
-        if (next_slot >= term_slot) {
-            pybind11_fail("initialize_multiphase_module_def: not enough space in slots");
-        }
-        slots[next_slot++] = {Py_mod_multiple_interpreters, detail::multi_interp_slot(options...)};
-#endif
-
-        if (detail::gil_not_used_option(options...)) {
-#if defined(Py_mod_gil) && defined(Py_GIL_DISABLED)
-            if (next_slot >= term_slot) {
-                pybind11_fail("initialize_multiphase_module_def: not enough space in slots");
-            }
-            slots[next_slot++] = {Py_mod_gil, Py_MOD_GIL_NOT_USED};
-#endif
-        }
-
-        // slots must have a zero end sentinel
-        if (next_slot > term_slot) {
-            pybind11_fail("initialize_multiphase_module_def: not enough space in slots");
-        }
-        slots[next_slot++] = {0, nullptr};
-
-        // module_def is PyModuleDef
-        // Placement new (not an allocation).
-        new (def) PyModuleDef{/* m_base */ PyModuleDef_HEAD_INIT,
-                              /* m_name */ name,
-                              /* m_doc */ options::show_user_defined_docstrings() ? doc : nullptr,
-                              /* m_size */ 0,
-                              /* m_methods */ nullptr,
-                              /* m_slots */ &slots[0],
-                              /* m_traverse */ nullptr,
-                              /* m_clear */ nullptr,
-                              /* m_free */ nullptr};
-        auto *m = PyModuleDef_Init(def);
-        if (m == nullptr) {
-            if (PyErr_Occurred()) {
-                throw error_already_set();
-            }
-            pybind11_fail("Internal error in module_::initialize_multiphase_module_def()");
-        }
```

Comment:
Sorry I forgot to ask before: IIUC, this code is dropped?

Is it better to simply return the `nullptr`? (Was the `error_already_set()` here actually ever correct?)

Reply:
throwing was probably wrong.  This is called from Python during import so really it should be setting a Python error (not throwing a C++ exception).